packages/hardhat-core/test/internal/hardhat-network/provider/modules/hardhat.ts

```diff
@@ -365,6 +374,650 @@ describe("Hardhat module", function () {
+            this.provider,
+            "hardhat_setStorageSlot",
+            [DEFAULT_ACCOUNTS_ADDRESSES[0].toString(), "xyz", intToHex(99)],
+            'Errors encountered in param 1: Invalid value "xyz" supplied to : QUANTITY'
+          );
+        });
+
+        it("should reject a storage key that is greater than 32 bytes", async function () {
+          const MAX_WORD_VALUE = new BN(2).pow(new BN(256));
+          await assertInvalidInputError(
+            this.provider,
+            "hardhat_setStorageSlot",
+            [
+              DEFAULT_ACCOUNTS_ADDRESSES[0].toString(),
+              numberToRpcQuantity(MAX_WORD_VALUE.add(new BN(1))),
+              "0xff",
+            ],
+            "Storage key must not be greater than 2^256"
+          );
+        });
+
+        for (const badInputLength of [1, 2, 31, 33, 64]) {
+          it(`should reject a value that is ${badInputLength} (not exactly 32) bytes long`, async function () {
+            await assertInvalidInputError(
+              this.provider,
+              "hardhat_setStorageSlot",
+              [
+                DEFAULT_ACCOUNTS_ADDRESSES[0].toString(),
+                intToHex(0),
+                `0x${"ff".repeat(badInputLength)}`,
+              ],
+              "Storage value must be exactly 32 bytes long"
+            );
+          });
+        }
+
+        it("should not reject valid argument types", async function () {
+          await this.provider.send("hardhat_setStorageSlot", [
+            DEFAULT_ACCOUNTS_ADDRESSES[0].toString(),
+            intToHex(0),
+            `0x${"ff".repeat(32)}`,
+          ]);
+        });
+
+        it("should result in modified storage", async function () {
+          const targetStorageSlot = 1;
+          const targetStorageValue = 99;
+          await this.provider.send("hardhat_setStorageSlot", [
+            DEFAULT_ACCOUNTS_ADDRESSES[0],
+            intToHex(0),
+            `0x${new BN(targetStorageValue).toString(16, 64)}`,
+          ]);
+
+          const resultingStorageValue = await this.provider.send(
+            "eth_getStorageAt",
+            [DEFAULT_ACCOUNTS_ADDRESSES[0], intToHex(0), "latest"]
+          );
+
+          assert.equal(resultingStorageValue, targetStorageValue);
+        });
+
+        it("should permit a contract call to read an updated storage slot value", async function () {
+          // Arrange: Deploy a contract that can get and set storage.
+          const [
+            ,
+            {
+              contracts: {
+                ["literal.sol"]: { Storage: storageContract },
+              },
+            },
+          ] = await compileLiteral(
+            `contract Storage {
+              function getValue(uint256 slot) public view returns (uint256 result) {
+                assembly { result := sload(slot) }
+              }
+              function setValue(uint256 slot, uint256 val) public {
+                assembly { sstore(slot, val) }
+              }
```

Comment:
The `setValue` function doesn't seem to be used, can we remove it?

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/1435/commits/8554a3e4a4b1a2d7ef0449e8cca93e029e270f4f